test/Autofac.Test/Core/ComponentRegistrationTests.cs

```diff
@@ -86,4 +86,71 @@ public void AttachingToPipelineBuildingShouldFailIfAlreadyBuilt()
 
         Assert.Throws<InvalidOperationException>(() => registration.PipelineBuilding += (s, p) => { });
     }
+
+    [Fact]
+    public void CanReplaceActivatorBeforePipelineHasBeenBuilt()
+    {
+        var services = new Service[] { new TypedService(typeof(object)) };
+
+        var registration = Factory.CreateSingletonRegistration(services, Factory.CreateProvidedInstanceActivator(new object()));
+
+        var newActivator = Factory.CreateProvidedInstanceActivator(new object());
+
+        registration.ReplaceActivator(newActivator);
+
+        Assert.Same(registration.Activator, newActivator);
+    }
+
+    [Fact]
+    public void CannotReplaceActivatorAfterRegistrationHasBeenBuilt()
+    {
+        var services = new Service[] { new TypedService(typeof(object)) };
+
+        var registration = Factory.CreateSingletonRegistration(services, Factory.CreateProvidedInstanceActivator(new object()));
+
+        var builder = Factory.CreateEmptyComponentRegistryBuilder();
+        builder.Register(registration);
+
+        builder.Build();
+
+        Assert.Throws<InvalidOperationException>(() => registration.ReplaceActivator(Factory.CreateProvidedInstanceActivator(new object())));
+    }
+
+    [Fact]
+    public void CannotReplaceActivatorIfTargetingAnotherRegistration()
+    {
+        var services = new Service[] { new TypedService(typeof(object)) };
+
+        var registration = Factory.CreateSingletonRegistration(services, Factory.CreateProvidedInstanceActivator(new object()));
+
+        var targetingRegistration = Factory.CreateTargetingRegistration(registration);
+
+        Assert.Throws<InvalidOperationException>(() => targetingRegistration.ReplaceActivator(Factory.CreateProvidedInstanceActivator(new object())));
+    }
+
+    [Fact]
+    public void ReplacementActivatorLimitTypeMustSatisfyAllServices()
+    {
+        var services = new Service[] { new TypedService(typeof(IServiceA)), new TypedService(typeof(IServiceB)) };
+
+        var registration = Factory.CreateSingletonRegistration(services, Factory.CreateProvidedInstanceActivator(new ComponentAB()));
+
+        Assert.Throws<ArgumentException>(() => registration.ReplaceActivator(Factory.CreateProvidedInstanceActivator(new ComponentA())));
+    }
+
```

Comment:
It may just be me being paranoid, but we use services for things like `Startable` that aren't typed. It might be worth adding a test to make sure `AutoActivate` on a non-`IStartable` item still works. Possibly also adapters/decorators, just to close the loop. I mean, I don't see any reason they shouldn't, but the weird edge cases freak me out. "I replaced the activator on my decorator type and now nothing works WHY DID YOU BREAK MY STUFF AUTOFAC?"

Reply:
Adding tests to `Specification.Test` for a variety of these features + activator replacement, and just wanted to note that you cannot replace the Activator for an Adapter, because it targets a different registration.  Only registrations that do not target others are permitted.

That restriction is mostly in there to stop people replacing the Activator for the `ExternalComponentRegistration` incorrectly, and to stop them implementing custom `Lazy<T>`, `Meta<T>` activators, but has also caused this problem with adapters (have I said how much I hate adapters before?).

Using `Targets` may be a bad approach though; it may be that I need to think of another way of preventing this, e.g. virtual property `CanReplaceActivator` or something that allows `ExternalComponentRegistration` to say "no".

One of the challenges I face is that the `Registered` event gets called for _all_ registrations added to child scopes, _including_ registrations from an external registry:

```csharp
public void ReplacementActivatorLimitCanReplaceInChildScope()
{
    var builder = new ContainerBuilder();

    builder.ComponentRegistryBuilder.Registered += (sender, args) =>
    {
        // I added a `ProvidesService` method to make this code less horrible.
        if (args.ComponentRegistration.ProvidesService<ServiceA>())
        {
            // This runs twice, fails the second time because the registration is an `ExternalComponentRegistration`.
            args.ComponentRegistration.ReplaceActivator(
                new CustomActivator<ServiceA>(_ => new ServiceA(createdCustom: true)));
        }
    };

    builder.RegisterType<ServiceA>();

    var container = builder.Build();

    var scope = container.BeginLifetimeScope(builder => { });

    var instance = scope.Resolve<ServiceA>();

    Assert.True(instance.CreatedCustom);
}
```

I can imagine this causing people some substantial confusion.  Should the `Registered` event even fire for `ExternalComponentRegistration` instances?

Decorators work fine because they don't create targeting registrations. Don't love the syntax required to extract the decorated service though, but it is there:

```csharp
builder.ComponentRegistryBuilder.Registered += (sender, args) =>
{
    if (args.ComponentRegistration.Activator.LimitType == typeof(ServiceADecorator))
    {
        // Don't love this syntax much, was sort of expecting the automatic 'inject parameters'
        // functionality in the ReflectionActivator to 'just work', but does not.
        args.ComponentRegistration.ReplaceActivator(
            new CustomActivator<ServiceADecorator>(ctxt => new ServiceADecorator(
                (IServiceA)ctxt.Parameters.OfType<TypedParameter>().First(x => x is TypedParameter t && t.Type == typeof(IServiceA)).Value,
                createdCustom: true)));
    }
};
```

Reply:
I'm not going to lie, if I step back from this whole thing and think, "Yeah, let's allow folks to replace the activators as part of an event for this edge case over here" I also kind of want to /headdesk because it seems like it's a lot of somewhat confusing work to support this thing that will be a total foot-gun to anyone trying to consume it outside of this one very specific edge case. I mean, maybe the answer really is a custom `RegisterMyWrappedType<T>` extension that then does the AOP or whatever else and registers the proxied type directly instead of trying to replace the activator. Or maybe the answer is to just use the right DynamicProxy decorators and not support some extra custom AOP thing. And, yeah, I'm sorry that means there's no global one-place-to-hook-into-literally-everything-except-the-stuff-that-doesn't-allow-it hook. I'd be OK not having that. It feels like Autofac-as-the-instrument-of-AOP is just as questionable as Autofac-as-the-instrument-of-orcehstrating-your-app-startup-logic.

Anyway, I digress, and I'm a little hangry at the moment.

Re: adapters... uh... maybe you don't get to replace the activator for an adapter and that's just a shortcoming of the first iteration of this feature. I'd be fine with that. (I'll take adapters over ACTNARS any day of the week.)

Re: should `ExternalComponentRegistration` fire `Registered`... I don't have a good answer for that. It seems like it's getting brought in from the external registry so it is kinda getting registered since it's from a different context and this is the first time this context has seen it. But I get that it's also kinda _not_ getting registered because nothing has really changed and it's simply bringing in something that already existed. I'd be afraid if we _didn't_ raise that event, then someone is going to want the `ExternalComponentBroughtInJustNow` event so they can do something with just these special registrations and I'm just not interested in having that. Maybe the `Registered` event needs more info in the args, like whether it's... uh... "inherited" or not. Is inherited the right word? You know what I mean.

Reply:
Also thinking about it some more now you're questioning it, I wonder if it wouldn't be better for this use-case to add custom middleware pre-activator to the registration, instead of actually changing the activator.

That means that you can choose to invoke the existing activator if you want by calling `next()` in the middleware, or, if you want to completely replace it, set `ctxt.Instance` and never call `next()`.

The current approach in the PR means you _cannot_ access the existing activator.

Fundamentally activators are just middleware, right? So then this may not even need code changes, since you can already add pipeline middleware to all registrations (e.g. log4net example).

Reply:
re:IsExternalRegistration on the Registered event, no superb way to do it to be honest, perhaps checking `registration is ExternalComponentRegistration`? That's internal though.

To be honest, while it may seem a little odd, I don't recall ever seeing an issue where someone complained that the event handlers were doubled up?

Reply:
I mean, if it can be done via middleware and all the information is there... 🤷 that's kinda what the middleware approach was all about, right? Maybe it's just a matter of documenting more examples of middleware usage, beyond the log4net one?

Reply:
I'll double-check my thought once I'm back at a keyboard, but yeah, maybe we can solve this with docs.